### PR TITLE
Custom events and engines

### PR DESCRIPTION
**Pull request by AntoinePrv**
This is work in progress for #217 

# User defined events
I'm opening the PR unfinished to discuss how to actually add user define events given that `Enum` are not mutable.
In this first commit, I swap the value of `ignite.engine.engine` with a new `Enum` in the function `add_event` from `ignite.contrib.event`.
However this does not work in the following case:

```python
from ignite.engine import Events
# or ignite.engine.engine
from ignite.contrib.events import add_events

add_events(...)
# Or other function that end up adding event
```
This is because now `Events` is in the namespace of the current module. The `Events` object will not be changed and the user cannot access the new events to add new handles (unless weirdly making imports).

I suggest three solutions:
  1. Move the `Enum` inside the `Engine` class. Swaping the `Enum` becomes a mutation of the class so works properly.
  2. Use `aenum` package to make `Enum` mutable (+1 package dependency)
  3. Change the `Enum` structure of `Events` for something mutable

Any opinion/  suggestion welcome.

# Contrib engine
I added a contrib TBPTT trainer to have a example of creating custom events along the way.

# TODO

- [x] Add unit tests
- [x] Find a solution to add new `Events` to the original `Enum`
- [ ] Add examples
- [x] Add reference to TBPTT in docstring

### CODE REVIEW

**Comment by vfdev-5** [comment]
to be honest, personnally i'm not so convinced about the way to change Events as you proposed. Need to think about more elegant way.

**Comment by AntoinePrv** [comment]
@vfdev-5 I cannot find the comments...

> Concerning Events, to be honest, personnally i'm not so convinced about the way to change Events as you proposed. Need to think about more elegant way.

I'm not either, hopping someone has a good solution.

Of course this need testing, I'll work on that.

**Comment by vfdev-5** [comment]
In github interface you should see something like that : 
screen shot 2018-07-28 at 21 54 38

**Comment by AntoinePrv** [comment]
@vfdev-5 found this on [StackOverflow](https://stackoverflow.com/questions/10248666/how-to-view-line-comments-in-github):
>If you're here because you've left comments, and can see your own comments, but others cannnot, then check for a yellow pending label on your comments. Others cannot see your comment until you've completed the review process and selected whether you want to request changes. Once you click through a full review, then others will see your comments.

**Comment by vfdev-5** [comment]
@AntoinePrv correct, forgot to submit the review

**Comment by vfdev-5** [comment]
> Find a solution to add new Events to the original Enum

Maybe we could tweak `Engine` to accepts some custom events and add time events into `ignite.contrib.events`.

File: ignite/contrib/engine.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
+# coding: utf-8
```
[the remark this answers is not in the document]
Comment (AntoinePrv) [reply ignite/contrib/engine.py]: Perhaps even `trainers` and leave `engine` for ignite core.
  - engine -> class
  - trainer -> specific instance

[the remark this answers is not in the document]
Comment (vfdev-5) [reply ignite/contrib/engine.py]: > Perhaps even trainers and leave engine for ignite core.

In `ignite.contrib.engines` we could later add evaluators too. Even in your case, do you need a supervised evalutor or native one is ok ?

[the remark this answers is not in the document]
Comment (AntoinePrv) [reply ignite/contrib/engine.py]: > we could later add evaluators too

That's a good point, even though we could have `ignite.contrib.evaluators` module.

In my case no, I'm doing without. There are cases in time series where you want to feed back your output into your input. This does not need to be defined outside of the model (i.e. could be done within the `forward` of `nn.Module` with a simple `if not self.training` because no optimization is done at this point) but if eventually someone write a code for that, it doesn't hurt to have it as an evaluator and not rewrite the code.

File: ignite/engine/engine.py [hunk 2]
```diff
@@ -12,13 +12,13 @@
...
 class Events(Enum):
     """Events that are fired by the :class:`ignite.engine.Engine` during execution"""
-    EPOCH_STARTED = "epoch_started"
-    EPOCH_COMPLETED = "epoch_completed"
-    STARTED = "started"
-    COMPLETED = "completed"
-    ITERATION_STARTED = "iteration_started"
-    ITERATION_COMPLETED = "iteration_completed"
-    EXCEPTION_RAISED = "exception_raised"
+    EPOCH_STARTED = auto()
```
[the remark this answers is not in the document]
Comment (AntoinePrv) [reply ignite/engine/engine.py]: Removed it.

**Comment by AntoinePrv** [comment]
As for the Events, I changed it by giving the opportunity to pass an optional `Enum` of events at engine initialization.

File: ignite/contrib/engines.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    device=None
+):
+    """
+    Factory function for creating a trainer for truncated backpropagation
+    through time supervised models.
+
+    Args:
+        model (`torch.nn.Module`): the model to train
+        optimizer (`torch.optim.Optimizer`): the optimizer to use
+        loss_fn (torch.nn loss function): the loss function to use
```
Comment (vfdev-5) [inline ignite/contrib/engines.py]: nit: maybe just a callable ?

File: ignite/engine/engine.py [hunk 4]
```diff
@@ -37,6 +37,9 @@ class Engine(object):
     Args:
         process_function (Callable): A function receiving a handle to the engine and the current batch
             in each iteration, and returns data to be stored in the engine's state
+        events (Enum, optional): Additional events that can be fired. Use this
```
Comment (vfdev-5) [inline ignite/engine/engine.py]: nit: maybe better  `events` -> `custom_events` ?

File: ignite/contrib/engines.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            loss_t = loss_fn(y_pred_t, y_t)
+            loss_t.backward()
+            optimizer.step()
+
+            # Setting state of engine for consistent behaviour
+            engine.state.output = loss_t.item()
+            loss_list.append(loss_t.item())
+
+            # Fire event for end of iteration
+            engine._fire_event(Tbptt_Events.TIME_ITERATION_STARTED)
```
Comment (vfdev-5) [inline ignite/contrib/engines.py]: nit: `TIME_ITERATION_STARTED` -> `TIME_ITERATION_COMPLETED`

**Comment by vfdev-5** [comment]
@AntoinePrv looks good:
* [x] Find a solution to add new Events to the original Enum 

cc @alykhantejani

File: ignite/contrib/engines.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    model,
+    optimizer,
+    loss_fn,
+    tbtt_step,
+    dim=0,
+    device=None
+):
+    """
+    Factory function for creating a trainer for truncated backpropagation
+    through time supervised models.
```
Comment (vfdev-5) [inline ignite/contrib/engines.py]: Could you please also add more context and link on the tbptt

**Comment by vfdev-5** [comment]
@AntoinePrv Looks good, just a flake fix needed

**Comment by AntoinePrv** [comment]
Adding the example is going to take a little more time.

File: ignite/contrib/engines.py [hunk 7]
```diff
@@ -36,10 +37,12 @@ def _detach_hidden(hidden):
     Auxillary function to cut the backpropagation graph by detaching the hidden
     vector.
     """
-    if isinstance(hidden, collections.Iterable):
+    if isinstance(hidden, torch.Tensor):
+        return hidden.detach()
+    elif isinstance(hidden, tuple):
```
Comment (vfdev-5) [inline ignite/contrib/engines.py]: nit: `elif isinstance(hidden, tuple):` -> `elif isinstance(hidden, (list, tuple)):`

Comment (AntoinePrv) [reply ignite/contrib/engines.py]: I removed `collections.Iterable` because it turns out that `Tensor` implements that (in an unexpected way).

In Pytorch, the hidden vectors for recurrent models are either a `Tensor` or a `tuple` of `Tensor`. I'll mimic the function you mentioned so that this can generalize to user-defined structures if needed.

File: ignite/engine/engine.py [hunk 8]
```diff
@@ -37,6 +37,9 @@ class Engine(object):
     Args:
         process_function (Callable): A function receiving a handle to the engine and the current batch
             in each iteration, and returns data to be stored in the engine's state
+        additional_events (Enum, optional): Additional events that can be fired.
```
Comment (vfdev-5) [inline ignite/engine/engine.py]: Just wonder whether it makes sense to have a list of custom events ? 
For example, I write another custom Tbptt trainer and want to use your Tbptt_Events plus another CustomEvent...
Maybe a solution will be to inherit of Tbptt_Events ...

**Comment by vfdev-5** [comment]
LGTM.

@alykhantejani @jasonkriss could you please take a look and give your opinion

I think we can agree on merge this without examples and provide an example in another PR.

File: tests/ignite/contrib/test_engines.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    assert not model.weight_hh_l0.item() == pytest.approx(0)
+
+
+def test_create_supervised_tbptt_trainer_with_cpu():
+    _test_create_supervised_tbptt_trainer("cpu")
+
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="Skip if no GPU")
+def test_create_supervised_tbptt_trainer_with_gpu():
+    _test_create_supervised_tbptt_trainer("gpu")
```
Comment (vfdev-5) [inline tests/ignite/contrib/test_engines.py]: nit: "gpu" -> "cuda"

Comment (vfdev-5) [reply tests/ignite/contrib/test_engines.py]: Actually, I can fix it by myself (seen a lot of times, but never tried and it works :)

**Comment by jasonkriss** [comment]
I have two minor discussion points...

1. Is it worth the complicated API to have event validation when attaching a handler? I wonder if it might be better to just allow firing/handling arbitrary events. New engines can implement their own event enums to make this easier but it wouldn't require changes to the core engine class. I know there's downsides to this, so I don't feel super strongly about it. Wanted to mention it as an option though.

2. If new engines are gonna be firing events, we should maybe consider changing `_fire_event` to `fire_event`.

**Comment by vfdev-5** [comment]
@jasonkriss thanks !
> If new engines are gonna be firing events, we should maybe consider changing _fire_event to fire_event. 

Nice catch. Yes, it should become public 

Concerning the first point, if I understand correctly your point, you propose to remove original event check:
```python
if event_name not in Events.__members__.values():
    self._logger.error("attempt to add event handler to an invalid event %s ", event_name)
    raise ValueError("Event {} is not a valid event for this Engine".format(event_name))
```
such that we can pass to `add_event_handler` any type object. I see your point it will much more flexible and without the overhead with `additional_events`, however the fact that there is not restriction on type of `event_name` makes me hesitating...
Maybe we could impose event_name to be a `str`...

**Comment by elanmart** [comment]
Hi, I also have two minor comments
1. Is there any reason why we want to fire custom `TIME_ITERATION_STARTED` in the propsed bptt-trainer? If this is only serving as an example, then perhaps we should put it into `examples`? 
2. @vfdev-5 how about having `accepted_events` as a property of the `Engine`? If you want to fire custom events, then maybe you should inherit `Engine`, and use something like.
```python
@property
def accepted_events(self):
    return super().accepted_events + [my, events, here]
```

**Comment by vfdev-5** [comment]
Hey @elanmart thanks for switching from pytorch and passing by with your comments :)

We discussed the reason to fire custom event [here](https://github.com/pytorch/ignite/issues/217#issuecomment-408243272) and idea here is to provide a contrib module with some specific trainers: an the first example tbptt trainer to solve however a list of specific tasks with RNNs and be reusable. That's why this is not in examples.

>  how about having accepted_events as a property of the Engine? If you want to fire custom events, then maybe you should inherit Engine, and use something like.

Sure that this can be a solution, but I agree with Jason Kriss that current `Engine` and event_name verification can be simplified without introducing additional attributes or inheritance...

What do you think ?

**Comment by jasonkriss** [comment]
@vfdev-5 Yea i think its ultimately a tradeoff between simplicity and safety. I can see an argument for each side. I subjectively would prefer the simpler side as I'd like to keep `Engine` as minimal as possible.

**Comment by elanmart** [comment]
@vfdev-5 I see, thanks! 

Well I'd probably vote for some form of validation (because I'm paranoid and often make stupid mistakes), but I'm certainly not a fan of having an extra arg in `Engine.__init__`.

**Comment by vfdev-5** [comment]
Okay, thanks you guys, let's wait for Aly's opinion. So, now we opt to remove `additional_events` and replace `event_name` type check by check if it is `str`:
```python
if not isinstance(event_name, str):
    self._logger.error("attempt to add event handler to an invalid event {} ".format(event_name))
    raise ValueError("Event name should be a string, but given {}".format(type(event_name)))
```

**Comment by alykhantejani** [comment]
@AntoinePrv thanks for the nice PR!

I have some comments around adding extra events. So I agree with @elanmart, I don't think we should add an extra arg to the constructor of the `Engine`. What about this:

Inside Engine's constructor we make a variable `self._allowed_events = list(Events.__members__.values())`

Then we have a method on `Engine`, `register_event(event_name)` which just does `self._allowed_events.append(event_name)`. This way `event_name` can be an enum value, a string or whatever. 

The check in `_fire_event` then becomes `if event_name in self._allowed_events:`

And finally, we should change `_fire_event` to `fire_event` (make it public)

**Comment by vfdev-5** [comment]
@alykhantejani really nice solution ! Thanks !

**Comment by AntoinePrv** [comment]
I made the discussed changes, which I think are the right direction. However, I think the interface to `fire_event` is still confusing:
  - IMHO, having it totally public can be confusing: a user not familiar with ignite may not know if he is responsible for firing the core events `ignite.engines.Events` (should it be allowed?)
  - The extra args passed to __all__ handlers (initially for handling the exception event I believe) are passed to all handlers.

File: ignite/engine/engine.py [hunk 10]
```diff
@@ -60,27 +57,58 @@ def train_and_store_loss(engine, batch):
...
-    def __init__(self, process_function, additional_events=None):
+    def __init__(self, process_function):
         self._event_handlers = defaultdict(list)
         self._logger = logging.getLogger(__name__ + "." + self.__class__.__name__)
         self._logger.addHandler(logging.NullHandler())
         self._process_function = process_function
         self.should_terminate = False
         self.should_terminate_single_epoch = False
         self.state = None
+        self._allowed_events = list()
```
Comment (vfdev-5) [inline ignite/engine/engine.py]: `list()` -> `[]`

**Comment by vfdev-5** [comment]
> IMHO, having it totally public can be confusing: a user not familiar with ignite may not know if he is responsible for firing the core events ignite.engines.Events (should it be allowed?)

@AntoinePrv I agree with this point. We can point out in the doc string the intended usage of the method, saying explicitly that "if no custom events in the engine, no need to call the method".

> The extra args passed to all handlers (initially for handling the exception event I believe) are passed to all handlers.

As you wrote in the docs, event *args, **kwargs are optional and its true that all handlers will receive them in case of a custom event.

So we have two types of args/kwargs:
- static = set during handler function setup
- dynamic = set during `fire_event` if case of custom events only

[Previously in the docs](https://pytorch.org/ignite/concepts.html#state) we suggested to use `engine.state` to exchange "dynamic" data between `process_function` and `handlers`.

To avoid double usage we can wrap `_fire_event` by a more user friendly `fire_event` without args/kwargs. 

What do you think ?

File: ignite/engine/engine.py [hunk 11]
```diff
@@ -212,17 +280,17 @@ def run(self, data, max_epochs=1):
         try:
             self._logger.info("Engine run starting with max_epochs={}".format(max_epochs))
             start_time = time.time()
-            self._fire_event(Events.STARTED)
+            self.fire_event(Events.STARTED)
```
[the remark this answers is not in the document]
Comment (AntoinePrv) [reply ignite/engine/engine.py]: To clarify: you say we  should leave `_fire_event` in all `Engine` internal calls?

[the remark this answers is not in the document]
Comment (vfdev-5) [reply ignite/engine/engine.py]: Yes, I think we could leave internally `self._fire_event`

File: ignite/engine/engine.py [hunk 12]
```diff
@@ -65,17 +65,50 @@ def __init__(self, process_function):
...
+        """
+        for name in event_names:
+            self._allowed_events.append(name)
+
     def add_event_handler(self, event_name, handler, *args, **kwargs):
         """Add an event handler to be executed when the specified event is fired
 
         Args:
-            event_name (Events): event from ignite.engine.Events to attach the handler to
+            event_name: event from ignite.engine.Events to attach the handler to
```
Comment (alykhantejani) [inline ignite/engine/engine.py]: should also add something like "or any event_name added by `register_events`"

File: ignite/engine/engine.py [hunk 13]
```diff
@@ -138,7 +171,7 @@ def on(self, event_name, *args, **kwargs):
         """Decorator shortcut for add_event_handler
 
         Args:
-            event_name (Events): event to attach the handler to
+            event_name: event to attach the handler to
```
Comment (vfdev-5) [inline ignite/engine/engine.py]: nit: same as said Aly on `add_event_handler`: 
event from `ignite.engine.Events` or any event_name added by `register_events` to attach the handler to

**Comment by vfdev-5** [comment]
@AntoinePrv thank you !

**Comment by vfdev-5** [comment]
@AntoinePrv FIY there is a discussion on TBPTT implementations [here](https://discuss.pytorch.org/t/implementing-truncated-backpropagation-through-time/15500/9).

**Comment by AntoinePrv** [comment]
Thanks for the heads up. I implemented on purpose  `k1 == k2` because it is much simpler and achieves most of the improvement.
I'm happy to work on a `k1 != k2` implementation, as well as providing the missing example in the near future.

**Comment by vfdev-5** [comment]
@AntoinePrv sounds good! So, send a PR when you have time to advance on this, as you know PR's are very welcome :)
